clarity-ext-scripts/clarity_ext_scripts/covid/import_samples.py

```diff
@@ -0,0 +1,105 @@
+import re
+import pandas as pd
+from clarity_ext.extensions import GeneralExtension
+from clarity_ext.utils import single
+from clarity_ext.domain import Container, Sample
+
+
+class Extension(GeneralExtension):
+    def is_control(self, name):
+        # Returns a control type if the name indicates that this is a control sample. Otherwise
+        # returns None.
+        for control_type, pattern in self.control_mapping.items():
+            if pattern.match(name):
+                return control_type
+        return None
+
+    def create_sample_or_control(self, original_name, project, index):
+        sample_name_fmt = self.context.current_step.udf_sample_name
+        control_name_fmt = self.context.current_step.udf_control_name
+
+        control_type = self.is_control(original_name)
+
+        name_fmt = control_name_fmt if control_type else sample_name_fmt
+        name = format_name(original_name, index, name_fmt, self.context.start)
+        sample_or_control = Sample(sample_id=None, name=name, project=project)
+        sample_or_control.udf_map.force(
+            "Control", "Yes" if control_type else "No")
+        if control_type:
+            sample_or_control.udf_map.force("Control type", control_type)
+        return sample_or_control
+
+    def execute(self):
+        self.control_mapping = dict()
+        for mapping in self.context.current_step.udf_control_mapping.split(","):
+            map_from, map_to = mapping.split(":")
+            self.control_mapping[map_to] = re.compile(map_from)
+
+        # 0. Get the project
+        project = self.context.clarity_service.get_project_by_name(
+            self.context.current_step.udf_project)
+
+        # 1. Create a 96 well plate in memory:
+        container_type = self.context.current_step.udf_container_type
+        container_name_fmt = self.context.current_step.udf_container_name
```

Comment:
I have two objections for not using the code already in place to generate container names. 1) Plate names in RNA and rtPCR steps assumes a naming convention for the initial plate name, namely a running number on second place of an underscore. This dependency is now hidden. 2) having existing code that claims to do the name generation, but is in fact not used is confusing.

Reply:
Right.

Just to make it clear: The script that's in place can't be used as it is, as it renames containers that already exist while this creates new containers.

It's hidden now, as you point out, because I wanted to allow users to set the container and sample name in the step to make it more powerful. This was mostly because we didn't have the requirements in place yet. (This pull request does not close the ticket yet, because we hadn't gotten the naming conventions yet.)

When this has been merged I was thinking of following up with a PR that sets the naming conventions discussed on the meeting today and then I can make it clear again, i.e. have the format string in the code. Sounds OK?

Reply:
What am I saying, I'll fix it now!